source/common/common/inline_map.h

```diff
@@ -0,0 +1,515 @@
+    template <class Key, class Value> friend class InlineMap;
+
+    /**
+     * Get the id of the inline entry in the inline array. This could be used to access the
+     * key/value pair in the inline map without hash searching.
+     */
+    uint32_t inlineId() const { return inline_id_; }
+
+    // This constructor should only be called by InlineMapDescriptor.
+    Handle(uint32_t inline_id) : inline_id_(inline_id) {}
+
+    const uint32_t inline_id_;
+  };
+
+  InlineMapDescriptor() = default;
+
+  /**
+   * Add an inline key and return related handle. If a same key is added more than once time the
+   * same handle will returned for the key. May only be called before finalize(). Heterogeneous
+   * lookup is supported here.
+   */
+  template <class GetKey> Handle addInlineKey(const GetKey& key) {
+    RELEASE_ASSERT(!finalized_, "Cannot create new inline key after finalize()");
+
+    if (auto it = inline_keys_map_.find(key); it != inline_keys_map_.end()) {
+      // If the key is already added, return related inline handle directly.
+      return it->second;
+    } else {
+      // If the key is not added yet, then create a new handle for this key.
+      ASSERT(inline_keys_map_.size() < std::numeric_limits<uint32_t>::max());
+      Handle handle{static_cast<uint32_t>(inline_keys_map_.size())};
+      inline_keys_.push_back(StorageKey(key));
+      inline_keys_map_.emplace(key, handle);
+      return handle;
+    }
+  }
+
+  /**
+   * Fetch the handle for the given key. Return handle if the given key is inline key and return
+   * absl::nullopt if the given key is normal key. May only be called after finalize(). This should
+   * be used to get the handle of the inline keys that added by addHandle(). This function could
+   * used to determine if a key is added as inline key or not at runtime or xDS config loading time
+   * and decide if the key should be used as inline key or normal key.
+   * Heterogeneous lookup is supported here.
+   */
+  template <class GetKey> absl::optional<Handle> getHandleByKey(const GetKey& key) const {
+    ASSERT(finalized_, "Cannot get inline handle before finalize()");
+
+    if (auto it = inline_keys_map_.find(key); it == inline_keys_map_.end()) {
+      return absl::nullopt;
+    } else {
+      return it->second;
+    }
+  }
+
+  /**
+   * Finalize this descriptor. No further changes are allowed after this point. This guarantees that
+   * all map created by the process have the same variable size and custom inline key adding. This
+   * method should only be called once.
+   */
+  void finalize() {
+    ASSERT(!finalized_, "Cannot finalize() an already finalized descriptor");
+    finalized_ = true;
+  }
+
+  /**
+   * @return true if the descriptor is finalized.
+   */
+  bool finalized() const { return finalized_; }
+
+  /**
+   * @return all inline keys that joined by the given separator.
+   */
+  std::string inlineKeysAsString(absl::string_view separator = ",") const {
```

Comment:
is this used anywhere?

Reply:
I think the whole lib is not used anywhere for now? But you can see how it works in the https://github.com/envoyproxy/envoy/pull/27928.

Reply:
Can you add a unit test for this one, though?

Reply:
> Can you add a unit test for this one, though?

done.